File: central-rbac-client/python/src/onelog_central_rbac/circuit_breaker.py
"""circuit_breaker.py — 3-state circuit breaker mirror @onelog/central-rbac-client.

closed → threshold failures → open → reset_sec → half-open → success → closed
                                                  half-open → failure → open

See ../../../SPEC.md §2.2 for state machine semantics.
"""
from __future__ import annotations

import time
from typing import Literal

State = Literal["closed", "open", "half-open"]


class CircuitBreaker:
    """3-state circuit breaker.

    Single-flight probe in half-open (prevents thundering herd).
    """

    def __init__(self, threshold: int, reset_sec: float) -> None:
        self._threshold = threshold
        self._reset_sec = reset_sec
        self._state: State = "closed"
        self._consecutive_failures = 0
        self._opened_at = 0.0
        self._probe_in_flight = False

    def can_proceed(self) -> bool:
        """Return True nếu request allowed. Manages state transitions."""
        if self._state == "closed":
            return True
        if self._state == "open":
            if (time.monotonic() - self._opened_at) >= self._reset_sec:
                self._state = "half-open"
                return self._claim_probe()
            return False
        # half-open
        return self._claim_probe()

    def _claim_probe(self) -> bool:
        if self._probe_in_flight:
            return False
        self._probe_in_flight = True
        return True

    def record_success(self) -> None:
        self._consecutive_failures = 0
        self._state = "closed"
        self._probe_in_flight = False

    def record_failure(self) -> None:
        self._probe_in_flight = False
        self._consecutive_failures += 1
        if self._state == "half-open":
            self._state = "open"
            self._opened_at = time.monotonic()
            return
        if self._consecutive_failures >= self._threshold:
            self._state = "open"
            self._opened_at = time.monotonic()

    def get_state(self) -> State:
        return self._state

File: central-rbac-client/nodejs/src/circuit-breaker.ts
/**
 * circuit-breaker.ts — 3-state circuit breaker (closed / open / half-open).
 *
 * Closed: normal operation, count failures.
 * Open:   threshold exceeded → reject requests immediately for resetSec.
 * Half-open: after resetSec, allow 1 probe request. Success → closed. Fail → open again.
 *
 * Single-flight probe: prevents thundering herd khi half-open state opens gate.
 */

type State = 'closed' | 'open' | 'half-open';

export class CircuitBreaker {
  private state: State = 'closed';
  private consecutiveFailures = 0;
  private openedAt = 0;
  private probeInFlight = false;

  constructor(
    private readonly threshold: number,
    private readonly resetMs: number,
  ) {}

  /** Check nếu can proceed. Returns true nếu allowed, false nếu circuit open. */
  canProceed(): boolean {
    if (this.state === 'closed') return true;
    if (this.state === 'open') {
      if (Date.now() - this.openedAt >= this.resetMs) {
        this.state = 'half-open';
        return this.claimProbe();
      }
      return false;
    }
    // half-open — only 1 probe at a time
    return this.claimProbe();
  }

  private claimProbe(): boolean {
    if (this.probeInFlight) return false;
    this.probeInFlight = true;
    return true;
  }

  /** Call sau khi request thành công. Reset state → closed. */
  recordSuccess(): void {
    this.consecutiveFailures = 0;
    this.state = 'closed';
    this.probeInFlight = false;
  }

  /** Call sau khi request fail. Increment counter, mở circuit nếu vượt threshold. */
  recordFailure(): void {
    this.probeInFlight = false;
    this.consecutiveFailures += 1;
    if (this.state === 'half-open') {
      // Half-open probe failed → back to open
      this.state = 'open';
      this.openedAt = Date.now();
      return;
    }
    if (this.consecutiveFailures >= this.threshold) {
      this.state = 'open';
      this.openedAt = Date.now();
    }
  }

  /** Return current state (for observability/testing). */
  getState(): State {
    return this.state;
  }
}
